File: src/claw_mem/security/validation.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Write Validator (MVP Version)

Validates memory write requests, rejects unsafe content.
"""

import re
from typing import List


class WriteValidator:
    """
    Write Validator
    
    MVP version uses rule-based matching. Semantic analysis will be added in future iterations.
    """
    
    # Unsafe patterns list (English + Chinese)
    UNSAFE_PATTERNS = [
        r"ignore.*instruction",  # Ignore instructions
        r"忽略.*指令",           # Ignore instructions
        r"override.*memory",     # Override memory
        r"覆盖.*记忆",           # Override memory
        r"delete.*file",         # Delete files
        r"删除.*文件",           # Delete files
        r"execute.*code",        # Execute code
        r"执行.*代码",           # Execute code
        r"system:",              # System prompt injection
        r"<\|startoftext\|>",   # Special tokens
        r"act as.*system",       # Role-playing attacks
        r"扮演.*系统",           # Role-playing attacks
    ]
    
    def __init__(self):
        """Initialize validator"""
        self.compiled_patterns = [
            re.compile(pattern, re.IGNORECASE)
            for pattern in self.UNSAFE_PATTERNS
        ]
    
    def validate(self, content: str) -> bool:
        """
        Validate memory content safety
        
        Args:
            content: Memory content
            
        Returns:
            bool: Safety status
        """
        # Check empty content
        if not content or len(content.strip()) == 0:
            return False
        
        # Check unsafe patterns
        for pattern in self.compiled_patterns:
            if pattern.search(content):
                return False
        
        # Check length (prevent overflow)
        if len(content) > 10000:
            return False
        
        return True
    
    def get_rejection_reason(self, content: str) -> str:
        """
        Get rejection reason
        
        Args:
            content: Memory content
            
        Returns:
            str: Rejection reason
        """
        if not content or len(content.strip()) == 0:
            return "Content is empty"
        
        for i, pattern in enumerate(self.compiled_patterns):
            if pattern.search(content):
                return f"Contains unsafe pattern: {self.UNSAFE_PATTERNS[i]}"
        
        if len(content) > 10000:
            return "Content too long (max 10000 characters)"
        
        return "Unknown reason"

File: src/security/validation.ts
// Copyright 2026 Peter Cheng
// Licensed under Apache-2.0

/**
 * Write Validator (MVP Version)
 *
 * Validates memory write requests, rejects unsafe content.
 * Uses rule-based pattern matching. Semantic analysis will be added
 * in future iterations.
 */

// Unsafe patterns list (English + Chinese)
const UNSAFE_PATTERNS: RegExp[] = [
  /ignore.*instruction/i,          // Ignore instructions
  /忽略.*指令/i,                    // Ignore instructions
  /override.*memory/i,              // Override memory
  /覆盖.*记忆/i,                     // Override memory
  /delete.*file/i,                  // Delete files
  /delete.*文件/i,                   // Delete files
  /execute.*code/i,                 // Execute code
  /执行.*代码/i,                     // Execute code
  /system:/i,                       // System prompt injection
  /<\|startoftext\|>/i,             // Special tokens
  /act as.*system/i,                // Role-playing attacks
  /扮演.*系统/i,                     // Role-playing attacks
];

const MAX_CONTENT_LENGTH = 10000;

export class WriteValidator {
  /**
   * Validate memory content safety.
   *
   * @param content - Memory content to validate
   * @returns `true` if content is safe, `false` otherwise
   */
  validate(content: string): boolean {
    // Check empty content
    if (!content || content.trim().length === 0) {
      return false;
    }

    // Check unsafe patterns
    for (const pattern of UNSAFE_PATTERNS) {
      if (pattern.test(content)) {
        return false;
      }
    }

    // Check length (prevent overflow)
    if (content.length > MAX_CONTENT_LENGTH) {
      return false;
    }

    return true;
  }

  /**
   * Get the rejection reason for content.
   *
   * @param content - Memory content to check
   * @returns Human-readable rejection reason
   */
  getRejectionReason(content: string): string {
    if (!content || content.trim().length === 0) {
      return "Content is empty";
    }

    for (let i = 0; i < UNSAFE_PATTERNS.length; i++) {
      if (UNSAFE_PATTERNS[i].test(content)) {
        return `Contains unsafe pattern: ${UNSAFE_PATTERNS[i].source}`;
      }
    }

    if (content.length > MAX_CONTENT_LENGTH) {
      return `Content too long (max ${MAX_CONTENT_LENGTH} characters)`;
    }

    return "Unknown reason";
  }
}
